### PR TITLE
Themes loader

### PR DESCRIPTION
**Pull request by zixxus**
### What are you changing/introducing
At first, this require:
https://github.com/zixxus/luya/tree/theme
https://github.com/zixxus/luya-env-dev/tree/themes
https://github.com/zixxus/luya-module-cms/tree/themes

Add Theme component
Add global variable @theme
Add folder /themes with XML file where we configure the template
Remove resources & views folder - now its be replaced

### What is the reason for changing/introducing
I decided that many people prefer to manage their templates in the same way as it is in wordpress or prestashop. In the coming days, the template management module will appear in the admin panel.

In the future, he also proposes to add child themes, which will help the developer sell one template to many clients.
Currently the default theme is / themes / blank. We can change them through the sql code: 

`SELECT * FROM `admin_config` WHERE `name` = 'active_theme' limit 1;`

### QA

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   |no

### CODE REVIEW

**Comment by boehsermoe** [comment]
Nice feature! I had thought of something like that too. At first please revert your filemode changes like `.gitignore 100644 → 100755` for more info see [here](https://stackoverflow.com/questions/1580596/how-do-i-make-git-ignore-file-mode-chmod-changes)
After that it is easier too review. Thank you!

**Comment by zixxus** [comment]
> Nice feature! I had thought of something like that too. At first please revert your filemode changes like `.gitignore 100644 → 100755` for more info see [here](https://stackoverflow.com/questions/1580596/how-do-i-make-git-ignore-file-mode-chmod-changes)
> After that it is easier too review. Thank you!

Thanks :) I change file mode, but I don't know i should create new pull request?

**Comment by nadar** [comment]
I will do a review tomorrow, this looks promising for me! i have some Inputs regarding naming, and i think we need an object for the theme itself, isntead of an array with keys.

is there a preview of the themes management module somewhere?

File: core/base/Boot.php [hunk 1]
```diff
@@ -246,6 +246,10 @@ public function applicationConsole()
             ],
         ]);
         $this->app = new ConsoleApplication($mergedConfig);
+
+        $this->theme = Yii::$app->themes->getTheme();
```
Comment (nadar) [inline core/base/Boot.php]: setting alias should be done in the bootstrap section /base/BaseBootstrap

File: core/base/Module.php [hunk 2]
```diff
@@ -178,6 +178,16 @@ public function getUrlRules()
      */
     public $moduleLayout = 'layout';
 
+    /**
+     * @var string The default file of layout
+     */
+    public $layout = 'main';
```
Comment (nadar) [inline core/base/Module.php]: layouts must be defined in the controller, i don't think we can introduce this variable here.

File: core/base/Module.php [hunk 3]
```diff
@@ -178,6 +178,16 @@ public function getUrlRules()
...
+
+    /**
+     * @var string The default folder of theme load
+     */
+    public $theme;
```
Comment (nadar) [inline core/base/Module.php]: The theme should be configured via the component, so i think the path for the layout should be extract from the themes component.

File: core/base/Module.php [hunk 4]
```diff
@@ -199,13 +209,23 @@ public function init()
...
      * the *@app* namespace views will be looked up for view files
      *
+     * Now use Themes component to load theme from database
+     *
      * @return string
      * @see \yii\base\Module::getLayoutPath()
      */
     public function getLayoutPath()
     {
+        $this->theme = Yii::$app->themes->getTheme();
```
Comment (nadar) [inline core/base/Module.php]: The layout file must be set in the controller, as mentioned above. Also not sure about error handling, cause there should be the question:

+ what kind of errors can a theme have?

File: core/components/Themes.php [hunk 5]
```diff
@@ -0,0 +1,188 @@
+<?php
```
Comment (nadar) [inline core/components/Themes.php]: In general, yes i agree we could added a themes component to luya which will be the main channel to communicate whether a theme exists or not, and which one is active. etc. So maybe we could keep this component, but improve it and make it the main channel to communicate between controller, views and application.

File: core/components/Themes.php [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+
+    /**
+     * Setup component
+     * @return $this
+     */
+    public function setup()
+    {
+
+        $this->config['active'] = $this->getActive();
```
Comment (nadar) [inline core/components/Themes.php]: instead of an array we need an object.

File: core/components/Themes.php [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+     *
+     * INSERT INTO `admin_config` (`name`, `value`, `is_system`, `id`) VALUES ('active_theme', '@app/themes/teamwant', '1', NULL);
+     *
+     * todo: create admin module to manage themes
+     * @return string
+     * @throws \yii\db\Exception
+     */
+    public function getActive()
+    {
+        try {
```
Comment (nadar) [inline core/components/Themes.php]: This won't work for other DBMS system therefore we would need active query, but its not allowed to make database requests in LUYA core as its mvc only and no database is needed. So the CMS must inject this information to the themes component in a bootstrap section for example.

File: tests/core/components/ThemesTest.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
+<?php
```
Comment (nadar) [inline tests/core/components/ThemesTest.php]: not sure how usefull those tests are :smile:

**Comment by nadar** [comment]
In addition i also would like to say that we maybe need an issue where we can discuss the GOALS of the themes option and how we can get there. What are specifications of a theme? how are they stored (xml? database?) how to install and load, etc.

**Comment by zixxus** [comment]
I review your comments and firstly I propose:

1) Rename all class and functions to Layouts (and similar to layouts)
Class Layouts {

init() - for initialize all script
getLayouts() - get all avaliable layouts (return all avaliable layouts, scan theme dir and load layout data from xml file)
getLayout() - get one layout by name (with layout data like name, descr, author etc. loaded from 
getActive() - get current "on" layout (only name)
getDir() - get Directory where layouts is stored.



}

2) Rewrite all arrays to objects.

2.1) Rewrite sql to Model.

2.2) 
> In general, yes i agree we could added a themes component to luya which will be the main channel to communicate whether a theme exists or not, and which one is active. etc. So maybe we could keep this component, but improve it and make it the main channel to communicate between controller, views and application.

I think it's a good idea, I will work on it.



3) Move code whose initialize layouts to CMS module and load it in controller (luya\cms\frontend\base\Controller).

In core i'll leave only Layouts component and also load this component.

In the administrator module I leave only Layouts managment wherein we can switch themes. Maybe I add function to create / load child themes but at a later time :)

Are you think I can split theme managment for two modules or it will be better when all code put into luya-module-cms/src/admin? In this I thinking about page in whose we can managment themes. This page will be placed for example in: https://demo.luya.io/en/admin#!/template/admin~2Flayouts~2Findex


4) Remove test class

5) Clear variables ( some variables can be place in component / controller).

98) Add theme uploader (.zip) and autoinstaller it.

99) Create docs for theme like what xml files should look like and etc.



----

> yes themes loading is a good idea, what about the yii framework way of this? pro/cons? 

Often I write themes and modules for wordpress, my clients like when page is simple, If he need to change theme or theme variables its be required in admin panel not in code. Also some clients buy themes and scipts and it need installer like "Theme install" button. I think it be more usefull than YII array with theme managment and more programmers will start using luya because it will be more readable.



----

What are you think about this? :)

**Comment by nadar** [comment]
@zixxus @boehsermoe i made an issue, so we can discuss there any maybe close this PR:

https://github.com/luyadev/luya/issues/1916

**Comment by zixxus** [comment]
> @zixxus @boehsermoe i made an issue, so we can discuss there any maybe close this PR:
> 
> #1916

OK :)

**Pull request closed without merging**